amalgamate.mjs

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env node
+
+import * as fsp from "fs/promises";
+import * as pathLib from "path";
+import * as child_process from "child_process";
+import { promisify } from "util";
+// import * as assert from "assert/strict";
+
+const execFile = promisify(child_process.execFile);
+
+const argv = process.argv.slice(2);
+const name = argv[0];
+
+const directory = process.cwd();
+
+const amalgamatedDir = "./amalgamated/";
+
+await fsp.mkdir(amalgamatedDir, { recursive: true });
+
+function amalgamatedPath(fileNameOrExtension) {
+    const fileName = fileNameOrExtension.startsWith(".") ? `${name}${fileNameOrExtension}` : fileNameOrExtension;
+    return pathLib.resolve(`${amalgamatedDir}/${fileName}`);
+}
+
+async function readLines(path) {
+    let buf;
+    try {
+        buf = await fsp.readFile(path);
+    } catch {
+        return [];
+    }
+    return buf
+        .toString()
+        .split("\n")
+        .filter(Boolean)
+        ;
+}
+
+const filesWithErrorsPath = amalgamatedPath("files_with_errors.txt");
+const filesWithErrors = new Set((await readLines(filesWithErrorsPath))
+    .map(path => pathLib.relative(".", path))
+);
+console.warn(`skipping ${filesWithErrors.size} files with errors`);
+await fsp.writeFile(filesWithErrorsPath, [...filesWithErrors].join("\n"));
+
+const flagsClangDoesntKnow = new Set([
+    // "-Wno-format-truncation",
+    // "-Wno-stringop-truncation",
+]);
+
+const skippedFiles = new Set([
+    // We use the real definitions in "psp/fsw/pc-linux/src/cfe_psp_start.c".
+    "osal/src/bsp/shared/src/bsp_default_app_startup.c",
+    "osal/src/bsp/shared/src/bsp_default_app_run.c",
+]);
+
+async function amalgamate({ redefinitions }) {
+    // console.log(redefinitions);
+
+    const commands = JSON.parse((await fsp.readFile("compile_commands.json")).toString())
+        .map(({ directory, file, output, command, arguments: args }) => {
+            const relativePath = pathLib.relative(".", file);
```

Comment:
If we enable the tests, we get duplicate files. We should check for that (e.g. see #2, or use whatever approach you think works).

Reply:
I don't remember exactly, but I think part of the problem is that even though there are duplicate files, they do different things (like the unit tests).  And since we don't care as much about the unit tests (at least I made that decision; if it's incorrect, we can switch things), I excluded them all.  If we just only choose the first of each file, then we may exclude the versions of a file that we actually want.